### PR TITLE
[XPU][CPU] Enable the native path of DeepSeek

### PR DESCRIPTION
**Pull request by airMeng**
## Motivation

Currently the native path of deepseek v3 is broken, so fix the broken native path which will benefit CPU and all other accelerators

Has verified on CPU and XPU:

```
TRITON_AVAILABLE=0 python3 -m sglang.bench_one_batch --batch-size 1 --input 32 --output 32 --model deepseek-ai/DeepSeek-R1 --trust-remote-code --device cpu --attention-backend torch_native --disable-radix --disable-mla
python3 -m sglang.bench_one_batch --batch-size 1 --input 32 --output 32 --model deepseek-ai/DeepSeek-R1 --trust-remote-code --device xpu --disable-mla
```

## Modifications

1. fix the broken path that sgl-kernel not build
2. let each quantization config determine whether the current HW is capable
3. enable the native FP8 path including MOE and quant
4. avoid the rotary embedding only available on CUDA

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by airMeng** [comment]
@chunyuan-w @YangQun1 Can you review CPU/HPU parts?

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 1]
```diff
@@ -51,7 +52,20 @@ def get_supported_act_dtypes(cls) -> List[torch.dtype]:
 
     @classmethod
     def get_min_capability(cls) -> int:
-        return 89  # Minimum hardware capability (e.g., Hopper GPUs).
+        if hasattr(torch, "cuda") and torch.cuda.is_available():
+            return 89
+
```
Comment (liangan1) [inline python/sglang/srt/layers/quantization/modelopt_quant.py:59]: Why the prior part use 12 for xpu?

Comment (airMeng) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:59]: In the theory, Gen12+ should support the advance dtype, is it true?

Comment (zhuyuhua-v) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:59]: xpu query capability for PVC will return 12.60.7

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1114,7 +1166,7 @@ def get_rope(
                 is_neox_style,
                 scaling_factor,
                 dtype,
-                **extra_kwargs,
+                device**extra_kwargs,
```
Comment (cboss6) [inline python/sglang/srt/layers/rotary_embedding.py]: Fix this typo.

**Review by cboss6 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/rotary_embedding.py (hunk 2 above)
Comment (airMeng) [reply python/sglang/srt/layers/rotary_embedding.py]: fixed

**Comment by zhaochenyang20** [comment]
@zhyncs who could review XPU?

**Comment by airMeng** [comment]
@zhaochenyang20 could you give a review?

**Comment by zhaochenyang20** [comment]
I don't know anything on XPU. Sorry.

**Comment by zhaochenyang20** [comment]
I will ask wangyang for help.

**Comment by yangw1234** [comment]
> I will ask wangyang for help.

Sure, I'll help review it. @zhaochenyang20

File: docs/start/install.md [hunk 3]
```diff
@@ -52,6 +52,15 @@ cd ..
 pip install -e "python[all_hip]"
 ```
 
+Note: To Intel GPU, do following instead:
+
+```
+git clone -b v0.4.2.post3 https://github.com/sgl-project/sglang.git
```
Comment (yangw1234) [inline docs/start/install.md]: Any reason why we use `v.4.2.post3` instead of `v0.4.3.post4` as cuda and ROCm?

File: python/sglang/srt/layers/quantization/base_config.py [hunk 4]
```diff
@@ -54,8 +54,18 @@ def get_min_capability(cls) -> int:
...
-        This requirement is due to the custom CUDA kernels used by the
-        quantization method.
+        This requirement is due to the custom kernels used by the
+        quantization method or the stock pytorch capability.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def get_available(cls) -> bool:
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/base_config.py]: Nit: is `is_available` or  `get_availability` more appropriate for this method?

File: python/sglang/srt/layers/quantization/fp8.py [hunk 5]
```diff
@@ -103,7 +110,24 @@ def get_supported_act_dtypes(cls) -> List[torch.dtype]:
...
     @classmethod
     def get_min_capability(cls) -> int:
-        return 80
+        if hasattr(torch, "cuda") and torch.cuda.is_available():
+            return 80
+        if hasattr(torch, "xpu") and torch.xpu.is_available():
+            return 0
+
+        # Vendors can update
+        return 999
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8.py]: 999 seems arbitrary. How about something like `sys.maxint`?

File: python/sglang/srt/layers/quantization/blockwise_int8.py [hunk 6]
```diff
@@ -72,7 +72,20 @@ def get_supported_act_dtypes(cls) -> List[torch.dtype]:
 
     @classmethod
     def get_min_capability(cls) -> int:
-        return 80
+        if hasattr(torch, "cuda") and torch.cuda.is_available():
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/blockwise_int8.py:75]: This is may not be useful scenario but what is the behavior if a user wants to use the `CPU` device on a server with `torch.cuda` available, e.g. he/she just wants to test some cpu feature?

File: python/sglang/srt/layers/quantization/fp8.py [hunk 7]
```diff
@@ -970,6 +1039,24 @@ def apply(
                 no_combine=no_combine,
             )
 
+        # for CPU and other accelerators, fallback to native path
+        return self.torch_w8a8_block_fp8_moe(
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8.py:1025]: `topk_weights`, `topk_ids` has already been computed above [here](https://github.com/sgl-project/sglang/pull/4086/files#diff-d88336bb1e74c8ee38a7f1c2258906dd0e6b6135d177ed83ea60d509a141264fR962-R972), are there any reasons that we cannot use them?

File: python/sglang/srt/layers/quantization/fp8.py [hunk 8]
```diff
@@ -870,6 +894,51 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
 
+    def torch_w8a8_block_fp8_moe(self, a, w1, w2, w1_s, w2_s, score, topk, block_shape):
+        from sglang.srt.layers.activation import SiluAndMul
+
+        """This function performs fused moe with block-wise quantization using native torch."""
+
+        B, D = a.shape
+        a = a.view(B, -1, D).repeat(1, topk, 1).reshape(-1, D)
+        out = torch.zeros(B * topk, w2.shape[1], dtype=a.dtype, device=a.device)
+        score = torch.softmax(score, dim=-1, dtype=torch.float32)
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8.py]: If I remember correctly, deepseek v3 does not use softmax gating method, have we tested the correctness on deepseek v3/r1?

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 9]
```diff
@@ -527,6 +566,61 @@ def get_w8a8_block_fp8_configs(
     return None
 
 
+def native_w8a8_block_fp8_matmul(A, B, As, Bs, block_size, output_dtype=torch.float16):
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8_kernel.py:728]: shall we add some unit tests for this function? e.g. comparing against the triton version.

**Review by yangw1234 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
@airMeng Hey, could resolve @yangw1234's ideas? Thanks!

File: python/sglang/srt/layers/quantization/fp8.py (hunk 8 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8.py]: fixed

File: python/sglang/srt/layers/quantization/fp8.py (hunk 7 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8.py:1025]: fixed in https://github.com/sgl-project/sglang/pull/4086/commits/372864aea808446fe47ec96e217007565160b430

File: python/sglang/srt/layers/quantization/fp8.py (hunk 5 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8.py]: fixed in https://github.com/sgl-project/sglang/pull/4086/commits/372864aea808446fe47ec96e217007565160b430

File: python/sglang/srt/layers/quantization/base_config.py (hunk 4 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/base_config.py]: fixed in https://github.com/sgl-project/sglang/pull/4086/commits/372864aea808446fe47ec96e217007565160b430

File: docs/start/install.md (hunk 3 above)
Comment (airMeng) [reply docs/start/install.md]: change to the latest main. Since we haven't got an stable release, we will suggest users to use the latest.

File: python/sglang/srt/layers/quantization/blockwise_int8.py (hunk 6 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/blockwise_int8.py:75]: There are a lot of ```is_cuda``` check anywhere else, the same as here.
For the scenario you mentioned, how about update ```def is_cuda``` https://github.com/sgl-project/sglang/blob/d052f4c8a9fb7e135ca0f0b09f6feead93db9e01/python/sglang/srt/utils.py#L78-L79 to
```python
def is_cuda():
    return hasattr(torch, "cuda") and torch.version.cuda is not None and get_bool_env_var("CUDA_AVAILABLE", default="true"):
```

If yes, I can update accordingly

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 9 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8_kernel.py:728]: in fact the function is exactly copied from https://github.com/sgl-project/sglang/blob/d052f4c8a9fb7e135ca0f0b09f6feead93db9e01/python/sglang/test/test_block_fp8.py#L89

Do you want me to keep the 2 copies as in the PR, or remove the copy in the test and import from here?

Comment (yangw1234) [reply python/sglang/srt/layers/quantization/fp8_kernel.py:728]: I think removing the copy in the test would make more sense? What do you think? @zhaochenyang20 @zhyncs

File: python/sglang/srt/layers/quantization/blockwise_int8.py (hunk 6 above)
Comment (yangw1234) [reply python/sglang/srt/layers/quantization/blockwise_int8.py:75]: I think maybe we can address this later.

File: python/sglang/srt/layers/quantization/fp8.py [hunk 10]
```diff
@@ -970,6 +1039,24 @@ def apply(
...
+            w1_s=(
+                layer.w13_weight_scale_inv
+                if self.block_quant
+                else layer.w13_weight_scale
+            ),
+            w2_s=(
+                layer.w2_weight_scale_inv if self.block_quant else layer.w2_weight_scale
+            ),
+            score=router_logits,
+            topk=top_k,
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8.py]: Also changing these arguments?

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 11]
```diff
@@ -124,6 +132,34 @@ def _per_token_group_quant_fp8_colmajor(
     tl.store(y_s_ptr, y_s)
 
 
+def native_per_token_group_quant_fp8(
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8_kernel.py:168]: maybe also remove copy in the test?

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 12]
```diff
@@ -194,19 +230,22 @@ def per_token_group_quant_fp8(
...
-            group_size,
-            N,
-            eps,
-            fp8_min=fp8_min,
-            fp8_max=fp8_max,
-            BLOCK=BLOCK,
-            num_warps=num_warps,
-            num_stages=num_stages,
-        )
+        if is_triton_:
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8_kernel.py]: @YangQun1 I remember HPU will also install the triton package but it does not support triton, right?

Comment (yangw1234) [reply python/sglang/srt/layers/quantization/fp8_kernel.py]: So maybe we also need to check device here.

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 13]
```diff
@@ -551,78 +645,82 @@ def w8a8_block_fp8_matmul(
...
-    M = A.numel() // A.shape[-1]
-
-    assert B.ndim == 2 and B.is_contiguous() and Bs.ndim == 2
-    N, K = B.shape
-    assert triton.cdiv(N, block_n) == Bs.shape[0]
-    assert triton.cdiv(K, block_k) == Bs.shape[1]
-
-    C_shape = A.shape[:-1] + (N,)
-    C = A.new_empty(C_shape, dtype=output_dtype)
+    if is_triton_:  # pragma: no cover
```
Comment (yangw1234) [inline python/sglang/srt/layers/quantization/fp8_kernel.py]: same as the `is_triton_` above

File: python/sglang/srt/model_loader/loader.py [hunk 14]
```diff
@@ -107,15 +107,25 @@ def _get_quantization_config(
         quant_config = get_quant_config(model_config, load_config)
         major, minor = get_device_capability()
 
-        if major is not None and minor is not None:
-            assert 0 <= minor < 10
-            capability = major * 10 + minor
-            if capability < quant_config.get_min_capability():
+        if not hasattr(quant_config, "get_available"):
```
Comment (yangw1234) [inline python/sglang/srt/model_loader/loader.py]: `get_availability`?

File: python/sglang/srt/utils.py [hunk 15]
```diff
@@ -105,6 +105,12 @@ def is_cuda_available():
     return torch.cuda.is_available() and torch.version.cuda
 
 
+def is_triton_available():
+    if not get_bool_env_var("TRITON_AVAILABLE", default="true"):
```
Comment (yangw1234) [inline python/sglang/srt/utils.py]: Do users need to set this env manually? Does checking device enough? e.g. if (`cuda` or `xpu`) and `has_triton_package`, then triton is available; otherwise triton is not available.

File: python/sglang/srt/layers/quantization/fp8.py (hunk 10 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8.py]: sorry, fixed

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 11 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8_kernel.py:168]: done

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 12 above)
Comment (airMeng) [reply python/sglang/srt/layers/quantization/fp8_kernel.py]: I have moved the device check into is_triton, is it okay?

File: python/sglang/srt/model_loader/loader.py (hunk 14 above)
Comment (airMeng) [reply python/sglang/srt/model_loader/loader.py]: sorry, fixed

File: python/sglang/srt/utils.py (hunk 15 above)
Comment (airMeng) [reply python/sglang/srt/utils.py]: I add the device check below.

As the mentioned above, HPU will also install the triton package but it does not support triton. I will leave here for HPU folks to update.   @YangQun1 any comment?

**Review by yangw1234 — APPROVED** [review]

**Comment by airMeng** [comment]
@zhaochenyang20 comments resolved, anything else to be addressed?

**Comment by yangw1234** [comment]
@zhaochenyang20 the code LGTM now.

**Comment by zhyncs** [comment]
@ispobock

**Comment by zhaochenyang20** [comment]
@zhyncs should we wait for ke?

**Comment by zhaochenyang20** [comment]
@airMeng Fix the lint plz.

**Comment by airMeng** [comment]
> @airMeng Fix the lint plz.

fixed

File: python/sglang/srt/layers/quantization/fp8_kernel.py (hunk 12 above)
Comment (YangQun1) [reply python/sglang/srt/layers/quantization/fp8_kernel.py]: yeah, HPU doesn't support trition.

File: python/sglang/srt/utils.py (hunk 15 above)
Comment (YangQun1) [reply python/sglang/srt/utils.py]: @airMeng could we move the device check before the package check? package can be installed by users, it may be hard to control.

File: python/sglang/srt/models/deepseek_v2.py [hunk 16]
```diff
@@ -646,19 +645,20 @@ def forward_absorb(
             )
         q_nope, q_pe = q.split([self.qk_nope_head_dim, self.qk_rope_head_dim], dim=-1)
 
-        if self.w_kc.dtype == torch.float8_e4m3fnuz:
+        if self.w_kc.dtype == torch.float8_e4m3fnuz or not is_cuda_available():
```
Comment (YangQun1) [inline python/sglang/srt/models/deepseek_v2.py]: why we need this device check here? torch.float8_e4m3fnuz is not hip specific?
```
fp8_type_ = torch.float8_e4m3fnuz if is_hip_ else torch.float8_e4m3fn
```

File: python/sglang/srt/utils.py (hunk 15 above)
Comment (airMeng) [reply python/sglang/srt/utils.py]: fixed @YangQun1

File: python/sglang/srt/models/deepseek_v2.py (hunk 16 above)
Comment (airMeng) [reply python/sglang/srt/models/deepseek_v2.py]: nice catch. fixed

**Comment by zhaochenyang20** [comment]
@yangw1234 could you review this again? We will merge this after your approval. Thanks1

**Comment by zhaochenyang20** [comment]
@airMeng Btw, could you change this title of the pr? I think it should be named with XPU/.

**Comment by yangw1234** [comment]
> @yangw1234 could you review this again? We will merge this after your approval. Thanks1

LGTM

**Comment by airMeng** [comment]
@zhaochenyang20 can you merge?

**Comment by zhaochenyang20** [comment]
sure! I will merge this after CI.

**Review by zhaochenyang20 — APPROVED** [review]

**Comment by airMeng** [comment]
@zhaochenyang20 can you help to trigger the CI?

**Comment by airMeng** [comment]
@merrymercy @zhyncs can we merge now since the CI passed on this commit https://github.com/sgl-project/sglang/pull/4086/commits/0379008a743a4937eb91bebc12cd746625e03a84 as we talked in the slack?
